### ISSUE TITLE
Filter Manager

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#39]
Create a central filter manager dialog or tab to list all applied or last filters, so users could enable/disable specific filters, last one or all of them. Currently user needs to go to each tab to disable its filter, that's annoying.

**Comment by tc-wleite** [comment]
Some users here asked for a button to clear all current filters, which I guess is related to the feature you proposed, although much simpler.

**Comment by lfcnassif** [comment]
Yes. I think keeping last filter is also useful, usually users don't want to list all case items in table, but see results of last filter applied.

**Comment by lfcnassif** [comment]
A pin filter feature should help a lot too.

**Comment by patrickdalla** [comment]
What would be the requirements for this?
There are many UI elements that add filters to result set. So one of the requirements, I imagine, is to show all these filters in a centralized location.
Additionally, 
1) this manager must implement some way to add other kind of filters?
1.1) What kind of additional filters? Query based I think would be one, and what others more? JavaScript based?
2) This manager must implement some way to save/persist these filters configuration?

**Comment by patrickdalla** [comment]
I have some thoughts about item 2 to discuss: One could say that when you create a bookmark with the filtered items, the filter doesn't need to be persisted any more. 

a) But saving the filters can help in understanding how that selection was made, beyond being useful to apply the same filter in other similar cases. 
b) This saved filter could be exported and configured to automatically create bookmarks in other cases at processing time.

We can break down 2.a) and 2.b) in other 2 enhancement issues, and let them to latter roadmap than 4.2.

**Comment by lfcnassif** [comment]
Hi @patrickdalla sorry for the delay here, I was busy with other things and I have a meeting in 5min...





> There are many UI elements that add filters to result set. So one of the requirements, I imagine, is to show all these filters in a centralized location.

Yes that is the idea. To avoid memory usage issues, we should use BitSets. For Multicases, we could use a Map<caseID, BitSet> to save the filters for each case, since the item IDs will be repeated. ItemID class uses a lot of memory and I think should be avoided.

1. I think the manager could be used just to store last filters results and reaply/disable them, not to create new ones, for that user can use the existing FilterCombobox (for those Query based) and other UI components. So Javascript filter definition would be another different feature...
2. If it is used just to store last filter results, as I thought initially, this wouldn't be a need, but we can store filter results to appear if user close and reopens the case, but I think this is not a must, but would help.
a) Hum, I see... But saving the Filter definition rule can be a bit difficult for some of them, like those from Metadata filtering panel, from the MapTab, GraphTab, maybe timeline...
b) I we manage to save the filter definition rule for all kind of filters (they can be even case specific if they use fields or values - From/To for exemple in Graph - that exists just in the case).

So I think we could start with just filter results management. Allow the user to apply set operations on the tracked BitSets results (union, intersection, subtraction) at first. Then we may improve to track the Filter definition rule later, what do you think?

**Comment by lfcnassif** [comment]
A non trivial point is: when user reapply the Filter results using the Manager, should it be reflected in the original panel used to create the Filter? Maybe that would need the Filter definition rule... Or maybe not, Filter Manager saved "filters" could be seen as another different filter (that remember last filters results) and could be combined with current Filters applied in other panels... Or should the UI workflow be totally refactored so applied Filters are shown just in the filter Manager and current UI filtering components stop being painted as RED to warn a filter is being applied? Any more possible UI interaction approaches?

**Comment by patrickdalla** [comment]
I think that restored filters should be reflected on the original panel.
I think that a some level of refactor is very useful and not so difficult.
We can define 2 upper level interfaces IFilterer(A UI component that creates filters) and IFilter. FilterManager would register IFilterer instances, which informs all defined filters.
When the saved filters state is restored, the corresponding filterers would be informed about the respective IFilter instances to be internally reflected also.

**Comment by patrickdalla** [comment]
As an initial stub: Would it be anything much different than this?

https://user-images.githubusercontent.com/28692427/223194332-82a3137e-aa4b-45c5-9a5f-36bac430a14b.mp4

**Comment by lfcnassif** [comment]
> Question: Should these differences be reflected to the user? Or should they be transparent?
I think that the description of the query filter should be the corresponding query text, what would already be a distinction point to the user.
I have not found any reason to evidence this difference in UI, but maybe you can think different.

I think they ideally could be transparent to the user, but shouldn't cause any warm if the difference is noticed.


> As an initial stub: Would it be anything much different than this?

Very cool! 

A feature that, in my opinion, would be very useful is allowing the user to specify the boolean operator to combine current filters. Today AND is always used, but user may want to use OR, or maybe do some custom combination of existing filters results.

**Comment by patrickdalla** [comment]
@lfcnassif , all of the filter panels does no expose the resulting bitset for now, right? So we need implement to make them expose these bitsets. Or am I missing something?

Althought for AND (interception) filtering bitsets weren't needed, they seems to be important for OR (union).

**Comment by lfcnassif** [comment]
> @lfcnassif , all of the filter panels does no expose the resulting bitset for now, right? So we need implement to make them expose these bitsets. Or am I missing something?

Yes, I think some of them are not exposing the results, those Query based are not for sure. Actually we don't use BitSets yet to store filter results, we use arrays of ItemId, but I think we should switch to BitSets (or SparseBitSets) to avoid excessive memory usage. Because ids can repeat in multicases, possible we would need a Map<CaseId, BitSet> to store results, this would need some refactorings...

As I understood from above video, we could save different filter results coming from the same Filterer panel, very nice! This was a feature I thought a long ago, but forgot to say, so we are having similar ideas :-)

Some open questions to all devs: should each filter result be automatically saved? Or should we limit to last X filters to avoid memory issues? Or perhaps do not save automatically and expose some "Save Filter" button?

**Comment by patrickdalla** [comment]
Let me know if I'm in the right way: Would it be the right sequence of filtering?

1) A UI Filterer element would "register" a filter.
1.1) That specific filter would be applied to Match All Docs query, to create the corresponding "bitset". 

2) When showing the resultset, it would be filtered against all registered and enabled filter bitsets

Is it?

**Comment by lfcnassif** [comment]
> Let me know if I'm in the right way: Would it be the right sequence of filtering?
> 
> 1. A UI Filterer element would "register" a filter.
>    1.1) That specific filter would be applied to Match All Docs query, to create the corresponding "bitset".
> 2. When showing the resultset, it would be filtered against all registered and enabled filter bitsets
> 
> Is it?

Makes sense. But I think it would change behavior at least of Metadata filter tab, since its results/filters today take into account just the result set currently being displayed, after all other filters was applied, not all case data (MatchAllDocsQuery). I have discussed this before with @tc-wleite in another issue, I think we didn't come up to a conclusion. I'll arrive at work after lunch and I can search for the discussion...

**Comment by patrickdalla** [comment]
Maybe, to avoid methods to "register" the filter, it could be somewhat diferent:

1) A UI Filterer element would create internally an filter.

2) When showing the resultset, it would be filtered against all cached and enabled filter bitsets
2.1) A specific filter would be applied to Match All Docs query, to create the corresponding "bitset", in its first usage.
2.2) This bitset would be saved in a cache, for latter reuse.

**Comment by lfcnassif** [comment]
Related issue: https://github.com/sepinf-inc/IPED/issues/1184

But with this feature, it won't be needed.

> I have discussed this before with @tc-wleite in another issue

I didn't find it, maybe it was a private conversation. The idea was to change Metadata Filter behavior so it would always behave globally, on all case items, not just on current result set after applying other filters. Aggregating filters BitSets using AND would give the same result on main table. But the values for the selected property and counts in Metadata filter panel would change if we change its behavior, I'm not sure what behavior is better, maybe both have pros/cons.

**Comment by patrickdalla** [comment]
In the filterer interface, I defined a method to return a global filter,
following its internal logic, and a getDefinedFilters, that would return
internal defined IFilter objects separately. For example, each selected
bookmark would be a separated filter returned by bookmarkfilterer.
In this second method the filterer should return the filters in an
independent way, so this filter could be placed in other container and
removed from the original filter.
So I created a container, that implements a filterer itself, that receives
these filter objects from other filterers positioned in the desired defined
tree logic rule (And , Or) Once they are 'copied' to this container, their
original can be removed from the source filterer, but would still be
available to this filterer.
So, for metadata panel, the filter would be defined based on the visible
resultset, but once copied to this filterer he would work independently.
Tomorow I will share a video to better explain.

Em qua., 8 de mar. de 2023 16:34, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by patrickdalla** [comment]
In this video I select a filter range suggested by MetadataPanel based on bookmark filtered resultset. Once I copy this filter to Structured filter, a query filter version is copied. So when I clear both metadata and bookmark panels, the range filter in structured filterer is applied to all the items.

https://user-images.githubusercontent.com/28692427/224019347-84e2c765-a981-4ff9-b656-a25dd8e1b5df.mp4

**Comment by patrickdalla** [comment]
In this other video I create and AND operator in "Structured Filterer" (anyone with better name?) with two bookmark filters.

https://user-images.githubusercontent.com/28692427/224020350-923389b8-32ac-4fb4-94f1-4911bf21efc4.mp4

**Comment by lfcnassif** [comment]
So current MetadataPanel filter behavior wouldn't need to change for this feature, nice ! (But we can discuss that change in a separate issue...)

The boolean tree is very nice! What about "Boolean Filter" or "Combined Filter" names? I also think a NOT node would be very useful to build combinations.

And should user be able to save the results of the filter tree as a new Filter? He can simply save the results as a new bookmark...

**Comment by patrickdalla** [comment]
> 

Hi @lfcnassif,

I am implementing the bitset cache filtering (SparseBitSet). I think that I can use the luceneId, making the cached bitsets exists in the App context, i. e., if you close a case and open it again, but in a multicase, all bitsets in cache will have to be recreated (instead of reloaded from some persisted location). 

I think you imagined to use a Map<CaseId, Bitset> because it would persists the results of the filter, and not the filter rule itself, right? Does exists any necessity to persist filter results? Do you see any other advantage of using this Map instead of bitsets based on luceneId?

**Comment by patrickdalla** [comment]
right

**Comment by patrickdalla** [comment]
Which is the appropriate method to get the Collection of Cases (or caseIDs)?

**Comment by patrickdalla** [comment]
> Which is the appropriate method to get the Collection of Cases (or caseIDs)?

I used it:

IPEDMultiSource appCase = App.get().appCase;
List<IPEDSource> cases = appCase.getAtomicSources();

It seems to work. Is it right?

**Comment by lfcnassif** [comment]
Yes, then you can call `IPEDSource.getSourceId()`. But be careful, if the Multicase cases loading order changes, that ID currently will change. That is not good, but should not affect this feature if results are not persisted.

**Comment by patrickdalla** [comment]
Hi @lfcnassif and @tc-wleite , I think the already implemented code can be transformed in a Pull Request. Just let me enumerate the features to you check if there is anything missing:

1) Filtermanager
- A filterer interface was defined, to represent all classes that defines filters that can be used by UI.
- An additional panel is shown with all filterers (named "applied filters").
- The following UI elements where adapted to this new Interface: 
![image](https://user-images.githubusercontent.com/28692427/226735855-988bdcf3-cde2-4d3a-a345-f23c6711349e.png)

- UICaseSearchFilter class was refactored to use the filters from these registered filterers (instead of 'hardcoded').
- A special CombinedFilterer was implemented. It is represented by a tree in the applied filters panel. Filters from other filterers can be dragged an dropped into it, to combine them in a new combined filter. 
- Once a filter is copied to this combined filter a background search on all docs with only this filter applied is done to create a bitset that will be put on a cache for later reuse.

2) Additionally, the metadata panel code was refactored to separate reusable code, in a way that it was possible to implement #1053  
- When clicking with right mouse button on a column header on main results table, a popup panel will be opened with the same results presented as if the corresponding metadata was selected on metadatapanel. This values can be selected to filter the resultset.

3) An additional popup menu was created to permit some filter options based on the corresponding cell value in results table. If the cell is multivalued, each one can be chosen in a submenu:
![image](https://user-images.githubusercontent.com/28692427/226739380-c4db4b76-36b4-4e83-b8b0-98a020941664.png)

**Comment by lfcnassif** [comment]
Awesome, thank you very much @patrickdalla! Just 2 quick questions:
1. Was the GraphTab updated to implement the new Filterer interface?
2. About (3) what does trigger the new menu instead the default context one?

**Comment by patrickdalla** [comment]
right mouse button with the CTRL key pressed triggers the cell value
context menu.
I forgot about graph. I will include it.

Em ter., 21 de mar. de 2023 17:23, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by patrickdalla** [comment]
Is it possible to create multiple CombinedFilterers?
Not in this implementation. We can do it or let as new issue. 

I think that soon it will lead to demands on saving the combined filters too. I was thinking in making filters serializable, but as soon as I saw that lucene queries aren't, I saw that it wouldn't be so easy. It would help to make clones of filters too, for copy operations between node on combined filters. For now, it is only possible to move filters between nodes in combined filters.

But let me know if you want me to improve this right now.
